### ISSUE TITLE
Create nodes required for mongodb sharding

### ISSUE THREAD

**Issue by jeff1evesque** [jeff1evesque/machine-learning#2924]
We need to implement our [chosen](https://github.com/jeff1evesque/machine-learning/issues/2842) puppet module, and create a basic mongodb shard cluster.

**Comment by jeff1evesque** [comment]
While creating this wrapper module, we made gained basic knowledge on mongodb from the following:

- http://www.codemag.com/Article/1309051
- https://forge.puppet.com/puppetlabs/mongodb

**Comment by jeff1evesque** [comment]
We need to remember to duplicate the puppet module for the `docker` environment. Additionally, we nest corresponding mongodb configuration under the `database` directive in our current `hiera.yaml`. This means, we'll need to move the current nested attributes under a new to-be created `database:mariadb` key, which means we'll need to adjust all manifests implementing these values.

**Comment by jeff1evesque** [comment]
We need to create a self-signed certificate, for each of our `mongodb.xxx` nodes.  Each of these CA's should have the same name as our corresponding yaml node file.  This means, we'll need to adjust our `Vagrantfile` respectively, so we can properly implement `"node/%{trusted.certname}.yaml"` path in our `hiera.yaml`.

**Comment by jeff1evesque** [comment]
Creating our certificates, should be a process similar to our [`install_puppet_agent`](https://github.com/jeff1evesque/puppet-demonstration/blob/master/install_scripts/install_puppet_agent#L92-L103) script. However, we'll need to tailor the syntax for `trusty64`, and segregate the logic into a bash script, contained in our `build/` directory`.  This will help reduce syntax, and keep logic succinct within our `Vagrantfile`.

**Comment by jeff1evesque** [comment]
As usualy, we need to ensure properl line endings for our bash scripts.  Otherwise, we may notice the typical line ending errors, in context during a `vagrant up` build:

```bash
...
==> aristotle: Running provisioner: shell...
    aristotle: Running: inline script
==> aristotle: stdin: is not a tty
==> aristotle: /tmp/vagrant-shell: ./configure-host: /bin/bash^M: bad interpreter: No such file or directory
==> aristotle: /tmp/vagrant-shell: ./configure-puppet: /bin/bash^M: bad interpreter: No such file or directory
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Note:** it may be worth noting that our `Vagrantfile` is essentially a ruby script.  This means at any location in the `Vagrantfile`, we may implement `puts some_variable`.  During our `vagrant up` build, we will see the value of `some_variable` within the terminal console.

**Comment by jeff1evesque** [comment]
It seems our implemented [mongodb](https://github.com/puppetlabs/puppetlabs-mongodb) module, is having some problems, during a `vagrant up` build:

```bash
$ vagrant up
...
==> aristotle: Running provisioner: puppet...
==> aristotle: Running Puppet with environment mongodb...
==> aristotle: stdin: is not a tty
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Stdlib::Compat::Hash. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/apt/manifests/init.pp", 18]:["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/mongodb/manifests/repo/apt.pp", 6]
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Stdlib::Compat::String. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/mongodb/manifests/server/config.pp", 88]:["/tmp/vagrant-puppet/modules-68f1d46533c7b79be8c41ff7869052d7/mongodb_cluster/manifests/init.pp", 7]
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Error: Evaluation Error: Error while evaluating a Function Call, 1 is not a string.  It looks to be a Fixnum at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/mongodb/manifests/server/config.pp:88:7 on node aristotle.mongodb.com
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by jeff1evesque** [comment]
If we comment out a segment of our `init.pp`, before rebuilding:

```puppet
class mongodb_cluster {
    contain mongodb_cluster::dependencies
#    contain mongodb_cluster::shard
    contain mongodb_cluster::databases
}
```

We receive the following adjusted traceback:

```bash
$ vagrant up
...
==> aristotle: Running provisioner: puppet...
==> aristotle: Running Puppet with environment mongodb...
==> aristotle: stdin: is not a tty
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Stdlib::Compat::Hash. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/apt/manifests/init.pp", 18]:["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/mongodb/manifests/repo/apt.pp", 6]
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Pattern[]. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/apt/manifests/setting.pp", 18]:
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Stdlib::Compat::Bool. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/apt/manifests/setting.pp", 19]:
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Warning: This method is deprecated, please use the stdlib validate_legacy function, with Stdlib::Compat::String. There is further documentation for validate_legacy function in the README. at ["/tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/apt/manifests/setting.pp", 37]:
==> aristotle:    (at /tmp/vagrant-puppet/modules-f07217ab419cc50c879fbe11d04e38c8/stdlib/lib/puppet/functions/deprecation.rb:25:in `deprecation')
==> aristotle: Notice: Compiled catalog for aristotle.mongodb.com in environment mongodb in 1.12 seconds
==> aristotle: Notice: /Stage[main]/Apt/File[preferences]/ensure: created
==> aristotle: Notice: /Stage[main]/Apt/Apt::Setting[conf-update-stamp]/File[/etc/apt/apt.conf.d/15update-stamp]/ensure: defined content as '{md5}0962d70c4ec78bbfa6f3544ae0c41974'
==> aristotle: Error: Failed to apply catalog: Could not evaluate MongoDB shell command: printjson(db.getMongo().getDBs())
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

So, we should proceed by isolating our larger problem, by iteratively ensuring a manifests functionality.

**Comment by jeff1evesque** [comment]
c0fb1b8: will properly install mongodb on the particular mongodb nodes. However, we still need to determine how to manually incorporate the following into custom puppet manifests:

- mongodb replication
- mongodb sharding
  - mongos
- create two databases

**Comment by jeff1evesque** [comment]
We need to increase the number of servers as follows:

- 2 mongodb shards
  - each shard is recommended to have 3 servers if replication is implemented
- 1 config servers as a replica set (CSRS)
  - 3 servers in replication
- 1 mongos
  - each instance of the application will have a corresponding mongos daemon

![mongodb](https://cloud.githubusercontent.com/assets/2907085/23936819/87f258fc-092a-11e7-8345-fd571f2710d3.png)

**Note:** it is acceptable for the config servers to have 1/5 the memory of implemented shard(s), and the cpu (in our case will be ignored) to be 1/2 of implemented shard(s).

**Comment by jeff1evesque** [comment]
We will not be creating the corresponding database(s) for this application in this issue.  Instead, it is more fitting to be taken care of via #2844.  This will allow us to focus more on setting up the necessary infrastructure pertaining to mongodb.

**Comment by jeff1evesque** [comment]
dddba6d: though we stated about a 1/5 ratio for memory on our configuration servers, we'll try the current values, and adjust in the future as needed.

**Comment by jeff1evesque** [comment]
807c3af: our `vagrant up` build would fail, at the end, when spinning up the config servers.  After increasing these config servers from `128` to `256`, every mongodb node successfully built.

**Comment by jeff1evesque** [comment]
ed51d0a: we are still trying to resolve our travis ci [unit test](https://travis-ci.org/jeff1evesque/machine-learning/builds/211565770#L5252-L5312) errors.

**Comment by jeff1evesque** [comment]
We need to determine the following cases when defining replica sets:

- should all mongodb servers be created before defining corresponding replica sets
- can mongodb servers can define replica sets, even if their sibling node has not yet been created

If all nodes need to be created before defining a replica set, we may need to add another provisioning step, after each mongodb servers have been created.

**Comment by jeff1evesque** [comment]
We need to determine the following cases when defining replica sets:

- should all mongodb servers be created before defining corresponding replica sets
- can mongodb servers define replica sets, even if their sibling node has not yet been created

If all nodes need to be created before defining a replica set, on a specific mongodb server, we may need to add another provisioning step, after each mongodb servers have been created.

**Comment by jeff1evesque** [comment]
Per mongodb manual for [deploying](https://docs.mongodb.com/manual/tutorial/deploy-replica-set/#considerations-when-deploying-a-replica-set) a replica set:

>The basic procedure is to start the [mongod](https://docs.mongodb.com/manual/reference/program/mongod/#bin.mongod) instances that will become members of the replica set, configure the replica set itself, and then add the [mongod](https://docs.mongodb.com/manual/reference/program/mongod/#bin.mongod) instances to it.
>...
>Before you can deploy a replica set, you must install MongoDB on each system that will be part of your [replica set](https://docs.mongodb.com/manual/reference/glossary/#term-replica-set). If you have not already installed MongoDB, see the [installation tutorials](https://docs.mongodb.com/manual/installation/#tutorial-installation).
>
>Before creating your replica set, you should verify that your network configuration allows communication among all members; i.e. each member must be able to connect to every other member. For instructions on how to check your connection, see [Test Connections Between all Members](https://docs.mongodb.com/manual/tutorial/troubleshoot-replica-sets/#replica-set-troubleshooting-check-connection).

Therefore, we will verify within each desired replica sets, that each vm can communicate to one another's private network.  When this has been verified, we will create a separate `configure_replicas.pp`, which will be responsible for creating replica sets.  This manifest will be `contain`ed in the `mongodb_cluster` module's `init.pp`.  However, `configure_replicas.pp` should have built in logic that checks if the proposed replica members can connect to each other, before enforcing the replica set.  If a native puppet syntax is not available, we could possibly enforce this via a combination of a bash script, controlled via an `onlyif` attribute to the `exec` directive.  Additionally, we will need to add an explicit provisioning step, which calls this `configure_replicas.pp`, after the mongodb nodes have been provisioned by `site.pp`.

**Comment by jeff1evesque** [comment]
After the necessary replia sets have been defined, we can proceed to create our sharded cluster as per successive mongodb documentation:

>This tutorial converts a single three-member replica set to a sharded cluster with two shards. Each shard is an independent three-member replica set. This tutorial is specific to MongoDB 3.4. For other versions of MongoDB, refer to the corresponding version of the MongoDB Manual.

>The procedure is as follows:

>1. Create the initial three-member replica set and insert data into a collection. See Set Up Initial [Replica Set](https://docs.mongodb.com/v3.4/tutorial/convert-replica-set-to-replicated-shard-cluster/#convert-setup-initial-set).
>2. Start the config servers and a [mongos](https://docs.mongodb.com/v3.4/reference/program/mongos/#bin.mongos). See [Deploy Config Server Replica Set and mongos](https://docs.mongodb.com/v3.4/tutorial/convert-replica-set-to-replicated-shard-cluster/#convert-deploy-sharding-infrastructure).
>3. Add the initial replica set as a shard. See [Add Initial Replica Set as a Shard](https://docs.mongodb.com/v3.4/tutorial/convert-replica-set-to-replicated-shard-cluster/#convert-add-initial-shard).
>4. Create a second shard and add to the cluster. See Add [Second Shard](https://docs.mongodb.com/v3.4/tutorial/convert-replica-set-to-replicated-shard-cluster/#convert-add-second-shard)
>5. Shard the desired collection. See [Shard a Collection](https://docs.mongodb.com/v3.4/tutorial/convert-replica-set-to-replicated-shard-cluster/#convert-shard-collection).

**Comment by jeff1evesque** [comment]
The following was run from `aristotle`:

```bash
vagrant@aristotle:~$ mongo --host 192.168.0.31 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.31:27017/test
> ^C
bye
vagrant@aristotle:~$ mongo --host 192.168.0.32 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.32:27017/test
```

The following was run from `socrates`:

```bash
vagrant@socrates:~$ mongo --host 192.168.0.30 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.30:27017/test
Thu Mar 16 18:37:33.711 Error: couldn't connect to server 192.168.0.30:27017 at src/mongo/shell/mongo.js:147
exception: connect failed
vagrant@socrates:~$ mongo --host 192.168.0.32 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.32:27017/test
Welcome to the MongoDB shell.
For interactive help, type "help".
For more comprehensive documentation, see
        http://docs.mongodb.org/
Questions? Try the support group
        http://groups.google.com/group/mongodb-user
> ^C
bye
```

The following was run from `plato`:

```bash
vagrant@plato:~$ mongo --host 192.168.0.30 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.30:27017/test
Thu Mar 16 18:40:46.348 Error: couldn't connect to server 192.168.0.30:27017 at src/mongo/shell/mongo.js:147
exception: connect failed
vagrant@plato:~$ mongo --host 192.168.0.32 --port 27017
MongoDB shell version: 2.4.9
connecting to: 192.168.0.32:27017/test
Thu Mar 16 18:41:25.973 Error: couldn't connect to server 192.168.0.32:27017 at src/mongo/shell/mongo.js:147
exception: connect failed
```

**Comment by jeff1evesque** [comment]
We destroyed, and rebuilt the above three environments.  It seems after a particular node has been accessed, via `mongo --host` command (from another node), it is no longer reachable by the same command.

**Comment by jeff1evesque** [comment]
This issue is getting too long.  We will ensure our mongo(s) is provisioned on the `main` application, then ensure both the application, and unit tests successfully builds.  When these are confirmed, and resolved, we will proceed to #2943 to continue provisioning the replicated shards.

**Comment by jeff1evesque** [comment]
We still have the following to readjust:

```bash
$ grep -r '/hiera/settings.yaml' .
Binary file ./.git/index matches
./brain/converter/crypto.py:        with open(root + "/hiera/settings.yaml", 'r') as stream:
./brain/converter/crypto.py:        with open(root + "/hiera/settings.yaml", 'r') as stream:
./brain/validator/validate_password.py:        with open(root + '/hiera/settings.yaml', 'r') as stream:
./brain/validator/validate_password.py:        with open(root + '/hiera/settings.yaml', 'r') as stream:
./factory.py:        prepath = 'hiera/' + args['prefix'] + '/hiera/settings.yaml'
./test/live_server/pytest.ini:#      'test/hiera/settings.yaml' configuration.
./test/manual/configure_database.py:    configuration = '/vagrant/hiera/test/hiera/settings.yaml'
./test/manual/pytest.ini:#      'test/hiera/settings.yaml' configuration.
```

**Comment by jeff1evesque** [comment]
Our application was able to successfully build, and run:

```bash
$ vagrant up
...
==> main: Notice: /Stage[main]/Nginx::Config/File[/var/nginx/client_body_temp]/ensure: created
==> main: Notice: /Stage[main]/Nginx::Config/File[/var/nginx/proxy_temp]/ensure: created
==> main: Notice: /Stage[main]/Nginx::Config/File[/etc/nginx/streams-enabled]/ensure: created
==> main: Notice: /Stage[main]/Nginx::Config/File[/etc/nginx/streams-available]/ensure: created
==> main: Notice: /Stage[main]/Nginx::Config/File[/etc/nginx/nginx.conf]/content: content changed '         {md5}a85dcbc5367783f3d42d0b7eab22ea50' to '{md5}de983fff1c3d7515a8100b89ed6ab532'
==> main: Notice: /Stage[main]/Webserver::Service/Nginx::Resource::Vhost[machine-learning.com]/Conc         at[/etc/nginx/sites-available/machine-learning.com.conf]/File[/etc/nginx/sites-available/machine-le         arning.com.conf]/ensure: defined content as '{md5}5937732bcc4cee3f211f266c2e3ccbb1'
==> main: Notice: /Stage[main]/Webserver::Service/Nginx::Resource::Vhost[machine-learning.com]/File         [machine-learning.com.conf symlink]/ensure: created
==> main: Notice: /Stage[main]/Nginx::Service/Service[nginx]: Triggered 'refresh' from 1 events
==> main: Notice: /Stage[main]/Webserver::Start/Service[start_gunicorn]/ensure: ensure changed 'sto         pped' to 'running'
==> main: Notice: Applied catalog in 791.22 seconds
```

![goto-results](https://cloud.githubusercontent.com/assets/2907085/24067942/8cfff67a-0b59-11e7-8bcd-883e1636f4c8.PNG)

![results](https://cloud.githubusercontent.com/assets/2907085/24067945/92b5c374-0b59-11e7-9c1e-290b8aa12a98.PNG)

**Note:** we still need to ensure our unit tests succeeds before we submit a pull request.

**Comment by jeff1evesque** [comment]
In this issue we decreased the memory for the `main` application from `10GB` to `6GB`, so we could allocate memory to our mongodb nodes without utilizing too much resources.  It seems our application was able to implement `scrypt` without any hiccups.  So, we may consider further decreasing the memory on the `main` application, when (or if) we consider [load balancing](https://github.com/jeff1evesque/machine-learning/issues/2937) our application.

**Comment by jeff1evesque** [comment]
f03feb5: we were able to run our current `main` application with `512MB` of memory.  While `scrypt` completed it's processes, it seemed noticeably slower (between 2-4x).  So, we'll generously reduce it to `1024MB`.  This should be sufficient for development, even later when we pair the main gunicorn application with a single `mongos` daemon.

5c44f0a: additionally, we removed non-pertinent docker stdout. This is not useful, since later we'll incorporate docker as a general practice, for our vagrant development environment.  Additionally, if a non-suitable unit test environment was the case, then our `live_server` unit tests would have failed.
